builder/builder-git.c

```diff
@@ -56,6 +56,19 @@ git_get_mirror_dir (const char     *url_or_path,
   g_autoptr(GFile) git_dir = NULL;
   g_autofree char *filename = NULL;
   g_autofree char *git_dir_path = NULL;
+  GPtrArray *sources_dirs = NULL;
+  int i;
+
+  sources_dirs = builder_context_get_sources_dirs (context);
+  for (i = 0; i < sources_dirs->len; i++)
+    {
+      GFile* sources_root = g_ptr_array_index (sources_dirs, i);
+      g_autoptr(GFile) git_dir = g_file_get_child (sources_root, "git");
```

Comment:
Yeah, I just tried to use a separate function for that cosntruction but at least my first approach does not look much nicer to me:

```

diff --git a/builder/builder-source-archive.c b/builder/builder-source-archive.c
index 8c71e9f..ff2bbbb 100644
--- a/builder/builder-source-archive.c
+++ b/builder/builder-source-archive.c
@@ -209,6 +209,16 @@ get_uri (BuilderSourceArchive *self,
 }
 
 static GFile *
+construct_download_path (BuilderSourceArchive *self,
+                         GFile *sources_root_dir,
+                         char *base_name)
+{
+  g_autoptr(GFile) sha256_dir = g_file_get_child (sources_root_dir, self->sha256);
+  g_autoptr(GFile) file = g_file_get_child (sha256_dir, base_name);
+  return g_steal_pointer (&file);
+}
+
+static GFile *
 get_download_location (BuilderSourceArchive *self,
                        BuilderContext       *context,
                        GError              **error)
@@ -216,8 +226,7 @@ get_download_location (BuilderSourceArchive *self,
   g_autoptr(SoupURI) uri = NULL;
   const char *path;
   g_autofree char *base_name = NULL;
-  GFile *download_dir = NULL;
-  g_autoptr(GFile) sha256_dir = NULL;
+  GFile *sources_root_dir = NULL;
   g_autoptr(GFile) file = NULL;
   GPtrArray *sources_dirs = NULL;
   int i;
@@ -239,17 +248,19 @@ get_download_location (BuilderSourceArchive *self,
   sources_dirs = builder_context_get_sources_dirs (context);
   for (i = 0; i < sources_dirs->len; i++)
     {
-      GFile* sources_root = g_ptr_array_index (sources_dirs, i);
-      g_autoptr(GFile) download_dir = g_file_get_child (sources_root, "downloads");
-      g_autoptr(GFile) sha256_dir = g_file_get_child (download_dir, self->sha256);
-      g_autoptr(GFile) file = g_file_get_child (sha256_dir, base_name);
-      if (g_file_query_exists (file, NULL))
+      GFile *root_dir = g_ptr_array_index (sources_dirs, i);
+      sources_root_dir = g_file_get_child (root_dir, "downloads");
+      file = construct_download_path (self, sources_root_dir, base_name);
+      if (g_file_query_exists (file, NULL)) {
+        g_clear_pointer (&sources_root_dir, g_object_unref);
         return g_steal_pointer (&file);
+      }
+      g_clear_pointer (&sources_root_dir, g_object_unref);
+      g_clear_pointer (&file, g_object_unref);
     }
 
-  download_dir = builder_context_get_download_dir (context);
-  sha256_dir = g_file_get_child (download_dir, self->sha256);
-  file = g_file_get_child (sha256_dir, base_name);
+  sources_root_dir = builder_context_get_download_dir (context);
+  file = construct_download_path (self, sources_root_dir, base_name);
 
   return g_steal_pointer (&file);
 }

```